File: src/main/java/org/stt/config/ConfigModule.java
package org.stt.config;


import dagger.Module;
import dagger.Provides;

import javax.inject.Named;
import java.io.File;

@Module
public class ConfigModule {
    private ConfigModule() {
    }

    @Provides
    static ConfigRoot provideConfigRoot(YamlConfigService configService) {
        return configService.getConfig();
    }

    @Provides
    static ActivitiesConfig provideTimeTrackingItemListConfig(ConfigRoot configRoot) {
        return configRoot.getActivities();
    }

    @Provides
    static WorktimeConfig provideWorktimeConfig(ConfigRoot configRoot) {
        return configRoot.getWorktime();
    }

    @Provides
    static ReportConfig provideReportConfig(ConfigRoot configRoot) {
        return configRoot.getReport();
    }

    @Provides
    static CliConfig provideCliConfig(ConfigRoot configRoot) {
        return configRoot.getCli();
    }

    @Provides
    static BackupConfig provideBackupConfig(ConfigRoot configRoot) {
        return configRoot.getBackup();
    }

    @Provides
    static CommonPrefixGrouperConfig provideCommonPrefixGrouperConfig(ConfigRoot configRoot) {
        return configRoot.getPrefixGrouper();
    }

    @Provides
    static JiraConfig provideJiraConfig(ConfigRoot configRoot) {
        return configRoot.getJira();
    }

    @Provides
    @Named("homePath")
    static String provideHomePath() {
        return determineBaseDir().getAbsolutePath();
    }

    private static File determineBaseDir() {
        String envHOMEVariable = System.getenv("HOME");
        if (envHOMEVariable != null) {
            File homeDirectory = new File(envHOMEVariable);
            if (homeDirectory.exists()) {
                return homeDirectory;
            }
        }
        return new File(System.getProperty("user.home"));
    }
}

File: src/main/kotlin/org/stt/config/ConfigModule.kt
package org.stt.config


import dagger.Module
import dagger.Provides
import java.io.File
import javax.inject.Named

@Module
class ConfigModule {

    @Provides
    fun provideConfigRoot(configService: ConfigServiceFacade): ConfigRoot {
        return configService.config
    }

    @Provides
    fun provideTimeTrackingItemListConfig(configRoot: ConfigRoot): ActivitiesConfig {
        return configRoot.activities
    }

    @Provides
    fun provideWorktimeConfig(configRoot: ConfigRoot): WorktimeConfig {
        return configRoot.worktime
    }

    @Provides
    fun provideReportConfig(configRoot: ConfigRoot): ReportConfig {
        return configRoot.report
    }

    @Provides
    fun provideCliConfig(configRoot: ConfigRoot): CliConfig {
        return configRoot.cli
    }

    @Provides
    fun provideBackupConfig(configRoot: ConfigRoot): BackupConfig {
        return configRoot.backup
    }

    @Provides
    fun provideCommonPrefixGrouperConfig(configRoot: ConfigRoot): CommonPrefixGrouperConfig {
        return configRoot.prefixGrouper
    }

    @Provides
    fun provideJiraConfig(configRoot: ConfigRoot): JiraConfig {
        return configRoot.jira
    }

    @Provides
    @Named("homePath")
    fun provideHomePath(): String {
        return determineBaseDir().absolutePath
    }

    private fun determineBaseDir(): File {
        val envHOMEVariable = System.getenv("HOME")
        if (envHOMEVariable != null) {
            val homeDirectory = File(envHOMEVariable)
            if (homeDirectory.exists()) {
                return homeDirectory
            }
        }
        return File(System.getProperty("user.home"))
    }
}
